### PR TITLE
[Site Isolation] Create a class to manage the aggregated set of activities on a WebContent process tree

### PR DESCRIPTION
**Pull request by pvollan**
#### 2cf2471c5f9fc902be3815f54a79069913a8a20f
<pre>
[Site Isolation] Create a class to manage the aggregated set of activities on a WebContent process tree
<a href="https://bugs.webkit.org/show_bug.cgi?id=307816">https://bugs.webkit.org/show_bug.cgi?id=307816</a>
<a href="https://rdar.apple.com/170326047">rdar://170326047</a>

Reviewed by Sihui Liu and Chris Dumez.

This change is introducing a new class, ProcessActivityGroup, that manages an aggregated set of activities
on a WebContent process tree. This ensures that the WebContent processes in the tree have the same activities.
When the ProcessActivityGroup is created, an activity is taken on the main frame process and all the iframe
processes. When a new iframe process is created, the same activity will be taken on it. And when an iframe
process is exiting, the associated activity will be destroyed. This can be used in suspension edge cases
where we currently are only taking an activity on the main frame process. This patch is covering some edge
cases. This is guarded behind the Site Isolation runtime flag.

New API test: SiteIsolation.ProcessActivityGroup

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::loadBackgroundWebView):
(WebKit::WebExtensionContext::loadInspectorBackgroundPage):
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
* Source/WebKit/UIProcess/RemotePageProxy.cpp:
(WebKit::RemotePageProxy::RemotePageProxy):
(WebKit::RemotePageProxy::~RemotePageProxy):
* Source/WebKit/UIProcess/SiteIsolatedActivity.cpp: Added.
(WebKit::SiteIsolatedActivity::create):
(WebKit::SiteIsolatedActivity::SiteIsolatedActivity):
(WebKit::SiteIsolatedActivity::~SiteIsolatedActivity):
(WebKit::SiteIsolatedActivity::createActivity):
(WebKit::SiteIsolatedActivity::addActivityForRemotePage):
(WebKit::SiteIsolatedActivity::removeActivityForRemotePage):
* Source/WebKit/UIProcess/SiteIsolatedActivity.h: Added.
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::didCreateRemotePage):
(WebKit::WebPageProxy::willDestroyRemotePage):
(WebKit::WebPageProxy::siteIsolatedActivity):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/UIProcess/XR/PlatformXRSystem.cpp:
(WebKit::PlatformXRSystem::ensureImmersiveSessionActivity):
* Source/WebKit/UIProcess/XR/PlatformXRSystem.h:

Canonical link: <a href="https://commits.webkit.org/309528@main">https://commits.webkit.org/309528@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/f4b89350eda64e4cee16b785b61477a974283fba

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/150726 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/23484 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/168/builds/17055 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/159448 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/104160 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/9e3e4f99-fedf-4da3-8a91-4b42e965e839/26b086bd-81f4-4456-951a-b77f0d57c516) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/152599 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/23916 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/23687 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/116333 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/82615 "Passed tests") | [loading 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/9e3e4f99-fedf-4da3-8a91-4b42e965e839/a4e9a44e-4893-4314-a075-12b8c465807b) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/153686 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/18445 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/135218 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/97061 "Passed tests") | | [loading 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/9e3e4f99-fedf-4da3-8a91-4b42e965e839/d791f939-428a-4e81-89bd-7975312bb0f1) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/17542 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/15489 "Passed tests") | [✅ 🛠 gtk3-libwebrtc](https://ews-build.webkit.org/#/builders/173/builds/7296 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/127156 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/167/builds/13142 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/161922 "Built successfully") | | 
| | [✅ 🛠 ios-safer-cpp](https://ews-build.webkit.org/#/builders/174/builds/5043 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/169/builds/14697 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/124332 "Passed tests") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/23286 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/19536 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/124530 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/23274 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/134937 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/79663 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/23190 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/19604 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/170/builds/11699 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/22888 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/86688 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/22600 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/22752 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/22654 "Built successfully") | | | |

### CODE REVIEW

File: Source/WebKit/UIProcess/XR/PlatformXRSystem.cpp [hunk 1]
```diff
@@ -91,10 +91,19 @@ void PlatformXRSystem::ensureImmersiveSessionActivity()
     if (!page)
         return;
 
-    if (m_immersiveSessionActivity && m_immersiveSessionActivity->isValid())
-        return;
+    bool siteIsolationEnabled = protect(page->preferences())->siteIsolationEnabled();
+    constexpr ASCIILiteral activityName = "XR immersive session"_s;
 
-    m_immersiveSessionActivity = protect(protect(page->legacyMainFrameProcess())->throttler())->foregroundActivity("XR immersive session"_s);
+    if (siteIsolationEnabled) {
```
Comment (szewai) [inline Source/WebKit/UIProcess/XR/PlatformXRSystem.cpp:103]: Maybe just early return in `siteIsolationEnabled` condition -- then we don't need `else`

File: Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm [hunk 2]
```diff
@@ -8082,4 +8083,17 @@ HTTPServer server({
...
+{
+    HTTPServer server({
+        { "/example"_s, { "<iframe src='https://apple.com/apple'></iframe>"_s } },
+        { "/apple"_s, { "hi"_s } }
+    }, HTTPServer::Protocol::HttpsProxy);
+
+    auto [webView, navigationDelegate] = siteIsolatedViewAndDelegate(server);
+    [webView.get().configuration _setClientNavigationsRunAtForegroundPriority:YES];
+    [webView loadRequest:[NSURLRequest requestWithURL:[NSURL URLWithString:@"https://example.com/example"]]];
+    [navigationDelegate waitForDidFinishNavigation];
```
Comment (szewai) [inline Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm]: What is this testing checking? Can we add testing only SPI to ensure the states are actually expected?

File: Source/WebKit/UIProcess/XR/PlatformXRSystem.cpp (hunk 1 above)
Comment (pvollan) [reply Source/WebKit/UIProcess/XR/PlatformXRSystem.cpp:103]: Good point, will fix!

Thanks for reviewing!

File: Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm (hunk 2 above)
Comment (pvollan) [reply Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm]: > What is this testing checking? Can we add testing only SPI to ensure the states are actually expected?

That's a great point. I have added this in the latest version of the patch.

The intention of this test was to check that the load finished when the Web view was hidden and had activities taken on the WebContent processes. However, I forgot to hide the Web view, but that is fixed now.

Thanks for reviewing!

File: Source/WebKit/WebKit.xcodeproj/project.pbxproj [hunk 3]
```diff
@@ -3,7 +3,7 @@
 	archiveVersion = 1;
 	classes = {
 	};
-	objectVersion = 55;
+	objectVersion = 54;
```
Comment (szewai) [inline Source/WebKit/WebKit.xcodeproj/project.pbxproj:6]: Needs to revert

File: Source/WebKit/UIProcess/ProcessActivityGroup.h [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#include "ProcessThrottler.h"
+#include "WebProcessProxy.h"
```
Comment (szewai) [inline Source/WebKit/UIProcess/ProcessActivityGroup.h:29]: This seems can be forward declared?

File: Source/WebKit/UIProcess/WebPageProxy.cpp [hunk 5]
```diff
@@ -17984,6 +17993,39 @@ void WebPageProxy::takeActivitiesOnRemotePage(RemotePageProxy& remotePage)
...
+    return internals();
+}
+
+Vector<RefPtr<WebProcessProxy>> WebPageProxy::Internals::activityTargets()
+{
+    Vector<RefPtr<WebProcessProxy>> targets;
+
+    Ref protectedPage = page;
+
+    RefPtr process = protectedPage->siteIsolatedProcess();
```
Comment (szewai) [inline Source/WebKit/UIProcess/WebPageProxy.cpp]: Why not using `Ref`?

File: Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm [hunk 6]
```diff
@@ -8308,4 +8310,28 @@ HTTPServer server({
...
+    HTTPServer server({
+        { "/example"_s, { "<iframe src='https://apple.com/apple'></iframe>"_s } },
+        { "/apple"_s, { "hello"_s } }
+    }, HTTPServer::Protocol::HttpsProxy);
+
+    auto [webView, navigationDelegate] = siteIsolatedViewAndDelegate(server, CGRectMake(0, 0, 800, 600));
+    bool finishedLoading { false };
+    navigationDelegate.get().didFinishNavigation = makeBlockPtr([&](WKWebView *, WKNavigation *navigation) {
+#if !PLATFORM(IOS_SIMULATOR)
+        // In the Simulator, we are never taking an activity when the view is hidden and we are calling
```
Comment (szewai) [inline Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:8406]: Could we set different expectations on `PLATFORM(IOS_SIMULATOR)` instead? 
```
unsigned expectedGroupSize= 2;
#if !PLATFORM(IOS_SIMULATOR)
    // In simulator...
    expectedGroupSize = 0;
#endif
EXPECT_EQ([navigation _processActivityGroupSizeForTesting], expectedGroupSize);
```

File: Source/WebKit/UIProcess/ProcessActivityGroup.cpp [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+}
+
+void ProcessActivityGroup::removeActivityForTarget(WebProcessProxy& target)
+{
+    m_activities.remove(target.coreProcessIdentifier());
+}
+
+void ProcessActivityGroup::activityTimedOut()
+{
+    m_activities.clear();
```
Comment (szewai) [inline Source/WebKit/UIProcess/ProcessActivityGroup.cpp:84]: Should we set a flag to indicate this group has reached timeout and should not longer receive new `createActivity` request? Otherwise it seems the lifetime could be extended outside timeout?

File: Source/WebKit/UIProcess/API/APIPageConfiguration.h [hunk 8]
```diff
@@ -587,6 +587,7 @@ class PageConfiguration : public ObjectImpl<Object::Type::PageConfiguration> {
         RetainPtr<ClassStructPtr> attachmentFileWrapperClass;
         std::optional<WTF::Vector<WTF::String>> additionalSupportedImageTypes;
         bool clientNavigationsRunAtForegroundPriority { true };
+        double clientNavigationActivityTimeout { 30 };
```
Comment (szewai) [inline Source/WebKit/UIProcess/API/APIPageConfiguration.h]: Where is this used?

**Review by szewai — APPROVED** [review]

File: Source/WebKit/UIProcess/API/APIPageConfiguration.h (hunk 8 above)
Comment (pvollan) [reply Source/WebKit/UIProcess/API/APIPageConfiguration.h]: Good catches! I will update the PR addressing these issues shortly.

Thanks for reviewing!

File: Source/WebKit/UIProcess/API/Cocoa/WKNavigation.mm [hunk 9]
```diff
@@ -60,6 +60,11 @@ - (BOOL)_isUserInitiated
     return _navigation->wasUserInitiated();
 }
 
+- (unsigned)_processActivityGroupSizeForTesting
+{
+    return Ref { *_navigation }->processActivityGroupSizeForTesting();
```
Comment (cdumez) [inline Source/WebKit/UIProcess/API/Cocoa/WKNavigation.mm]: We normally use `protect()` in new code rather than `Ref { }`

File: Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm [hunk 10]
```diff
@@ -2999,16 +3005,23 @@ void effectiveAppearanceDidChange(Inspector::ExtensionAppearance appearance) ove
         Ref inspectorExtension = result.value();
         inspectorExtension->setClient(makeUniqueRef<InspectorExtensionClient>(inspectorExtension, *this));
 
+        // Use foreground activity to keep background content responsive to events.
         Ref process = inspectorBackgroundWebView._page->legacyMainFrameProcess();
```
Comment (cdumez) [inline Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm]: If you're going to introduce a `inspectorPage` local variable below then I'd swap the two local variable and use `Ref process = inspectorPage->legacyMainFrameProcess();`. We might as well.

File: Source/WebKit/UIProcess/ProcessActivityGroup.h [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+class ProcessActivityGroup;
+
+class ProcessActivityGroupContext
+: public CanMakeWeakPtr<ProcessActivityGroupContext>
+, public CanMakeCheckedPtr<ProcessActivityGroupContext> {
+    WTF_MAKE_TZONE_ALLOCATED(ProcessActivityGroupContext);
+    WTF_MAKE_NONCOPYABLE(ProcessActivityGroupContext);
+    WTF_OVERRIDE_DELETE_FOR_CHECKED_PTR(ProcessActivityGroupContext);
+public:
+    Ref<ProcessActivityGroup> foregroundProcessActivityGroup(ASCIILiteral name, std::optional<Seconds> timeout = std::nullopt);
```
Comment (cdumez) [inline Source/WebKit/UIProcess/ProcessActivityGroup.h]: Should probably come after the constructor/destructor for clarity.

File: Source/WebKit/UIProcess/ProcessActivityGroup.cpp [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+
+ProcessActivityGroup::~ProcessActivityGroup()
+{
+    if (CheckedPtr context = m_context.get())
+        context->removeProcessActivityGroup(*this);
+}
+
+Ref<ProcessThrottler::Activity> ProcessActivityGroup::createActivity(WebProcessProxy& process)
+{
+    Ref protectedProcess = process;
```
Comment (cdumez) [inline Source/WebKit/UIProcess/ProcessActivityGroup.cpp]: This doesn't make sense:
1. You never need to protect parameters given to you
2. We no longer use `protected` in variable names (except for protectedThis in lambdas).

File: Source/WebKit/UIProcess/ProcessActivityGroup.cpp [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+    return protect(protectedProcess->throttler())->backgroundActivity(m_name);
+}
+
+void ProcessActivityGroup::addActivityForTarget(WebProcessProxy& target)
+{
+    if (m_timeout.has_value() && !m_timer.isActive()) {
+        // The timer has fired, so we should not add new activities.
+        return;
+    }
+    Ref<ProcessThrottler::Activity> activity = createActivity(target);
```
Comment (cdumez) [inline Source/WebKit/UIProcess/ProcessActivityGroup.cpp]: Why is `<ProcessThrottler::Activity>` needed?

File: Source/WebKit/UIProcess/ProcessActivityGroup.cpp [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+    m_activities.set(target.coreProcessIdentifier(), activity);
```
Comment (cdumez) [inline Source/WebKit/UIProcess/ProcessActivityGroup.cpp]: `WTF::move(activity)` if you insist on using a local variable for `activity`

File: Source/WebKit/UIProcess/ProcessActivityGroup.cpp [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+}
+
+void ProcessActivityGroupContext::removeProcessActivityGroup(ProcessActivityGroup& activityGroup)
+{
+    m_processActivityGroups.remove(activityGroup);
+}
+
+void ProcessActivityGroupContext::didAddActivityTarget(WebProcessProxy& target)
+{
+    m_processActivityGroups.forEach([&](Ref<ProcessActivityGroup> activityGroup) {
```
Comment (cdumez) [inline Source/WebKit/UIProcess/ProcessActivityGroup.cpp]: `auto&`

File: Source/WebKit/UIProcess/ProcessActivityGroup.cpp [hunk 16]
```diff
@@ -0,0 +1,118 @@
...
+        activityGroup->addActivityForTarget(target);
+    });
+}
+
+void ProcessActivityGroupContext::didRemoveActivityTarget(WebProcessProxy& target)
+{
+    m_processActivityGroups.forEach([&](Ref<ProcessActivityGroup> activityGroup) {
```
Comment (cdumez) [inline Source/WebKit/UIProcess/ProcessActivityGroup.cpp]: `auto&`

File: Source/WebKit/UIProcess/WebPageProxy.cpp [hunk 17]
```diff
@@ -18001,6 +18010,39 @@ void WebPageProxy::takeActivitiesOnRemotePage(RemotePageProxy& remotePage)
...
+    return internals();
+}
+
+Vector<RefPtr<WebProcessProxy>> WebPageProxy::Internals::activityTargets()
+{
+    Vector<RefPtr<WebProcessProxy>> targets;
+
+    Ref protectedPage = page;
+
+    Ref process = protectedPage->siteIsolatedProcess();
```
Comment (cdumez) [inline Source/WebKit/UIProcess/WebPageProxy.cpp]: It doesn't seem like we need this local variable.

File: Source/WebKit/UIProcess/WebPageProxy.cpp [hunk 18]
```diff
@@ -18001,6 +18010,39 @@ void WebPageProxy::takeActivitiesOnRemotePage(RemotePageProxy& remotePage)
...
+    targets.append(process);
+
+    Ref group = protectedPage->browsingContextGroup();
```
Comment (cdumez) [inline Source/WebKit/UIProcess/WebPageProxy.cpp]: Doesn't seem like we need this local variable, it is only used once. We can just use `protect()`.

File: Source/WebKit/UIProcess/WebPageProxy.cpp [hunk 19]
```diff
@@ -18001,6 +18010,39 @@ void WebPageProxy::takeActivitiesOnRemotePage(RemotePageProxy& remotePage)
...
+    group->forEachRemotePage(protectedPage, [&](auto& remotePageProxy) {
+        RefPtr process = remotePageProxy.process();
```
Comment (cdumez) [inline Source/WebKit/UIProcess/WebPageProxy.cpp]: We don't need this local variable.

File: Source/WebKit/UIProcess/WebPageProxyInternals.h [hunk 20]
```diff
@@ -533,6 +536,8 @@ struct WebPageProxy::Internals final : WebPopupMenuProxy::Client
     RetainPtr<CocoaView> platformView() const final;
 #endif
 
+    Vector<RefPtr<WebProcessProxy>> activityTargets() final;
```
Comment (cdumez) [inline Source/WebKit/UIProcess/WebPageProxyInternals.h]: Why are we using `RefPtr` and not `Ref`?

Comment (pvollan) [reply Source/WebKit/UIProcess/WebPageProxyInternals.h]: Ah, good point! I will address these issues shortly.

Thanks for reviewing!

File: Source/WebKit/UIProcess/ProcessActivityGroup.cpp (hunk 13 above)
Comment (pvollan) [reply Source/WebKit/UIProcess/ProcessActivityGroup.cpp]: Good catch! In the latest patch, I changed this to just using `Ref`.

File: Source/WebKit/UIProcess/ProcessActivityGroup.cpp (hunk 12 above)
Comment (pvollan) [reply Source/WebKit/UIProcess/ProcessActivityGroup.cpp]: Ah, good point! This has been resolved in the latest patch.

File: Source/WebKit/UIProcess/ProcessActivityGroup.h (hunk 4 above)
Comment (pvollan) [reply Source/WebKit/UIProcess/ProcessActivityGroup.h:29]: I went back to including this file, due to a GTK build issue.

**Review by cdumez — APPROVED** [review]
